data_model/derive/src/event_set.rs

```diff
@@ -0,0 +1,395 @@
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let Self {
+            vis,
+            event_enum_ident,
+            set_ident,
+            variants,
+        } = self;
+
+        // definitions of consts for each event
+        let flag_defs = variants.iter().zip(0u32..).map(
+            |(
+                EventSetVariant {
+                    flag_ident,
+                    event_ident,
+                    ..
+                },
+                i,
+            )| {
+                let doc = format!(" Matches [`{event_enum_ident}::{event_ident}`]");
+                quote! {
+                    #[doc = #doc]
+                    #vis const #flag_ident: Self = Self(1 << #i);
+                }
+            },
+        );
+        // identifiers of all the flag constants to use in impls
+        let flag_idents = variants
+            .iter()
+            .map(|EventSetVariant { flag_ident, .. }| quote!(#set_ident::#flag_ident))
+            .collect::<Vec<_>>();
+        // names of all the flag (as string literals) to use in debug impl
+        let flag_names = variants
+            .iter()
+            .map(|EventSetVariant { flag_ident, .. }| {
+                let flag_name = flag_ident.to_string();
+                quote! {
+                    #flag_name
+                }
+            })
+            .collect::<Vec<_>>();
+        // names of all the flags in `quotes` to use in error message
+        let flag_names_str = variants
+            .iter()
+            .map(|EventSetVariant { flag_ident, .. }| format!("`{flag_ident}`"))
+            .collect::<Vec<_>>()
+            .join(", ");
+        // patterns for matching events in the `matches` method
+        let event_patterns = variants.iter().map(
+            |EventSetVariant {
+                 event_ident,
+                 fields_style,
+                 ..
+             }| {
+                quote! {
+                    #event_enum_ident::#event_ident #fields_style
+                }
+            },
+        );
+
+        let doc = format!(" An event set for [`{event_enum_ident}`]s\n\nEvent sets of the same type can be combined with a custom `|` operator");
+
+        tokens.extend(quote! {
+            #[derive(
+                Copy,
+                Clone,
+                PartialEq,
+                Eq,
+                PartialOrd,
+                Ord,
+                Hash,
+                // this introduces tight coupling with these crates
+                // but it's the easiest way to make sure those traits are implemented
+                parity_scale_codec::Decode,
+                parity_scale_codec::Encode,
+                // TODO: we probably want to represent the bit values for each variant in the schema
+                iroha_schema::IntoSchema,
+            )]
+            #[repr(transparent)]
+            #[doc = #doc]
+            #vis struct #set_ident(u32);
```

Comment:
can't we keep it private?

Reply:
It is supposed to be used by the client sdk users directly, so I guess not? (think of `AccountEventFilter::new().for_events(AccountEventSet::Created)`) Having control of exposed API was the main reason I opted to develop out own derive macro instead of relying on an existing one

Reply:
makes sense